### PR TITLE
dnf4: init at 4.18.1

### LINKED ISSUE #107428: cppunit: enabling tests in libdnf leads to "Undefined symbols for architecture x86_64"
**Describe the bug**
As seen in https://github.com/NixOS/nixpkgs/pull/107348#discussion_r547528819 , trying to enable tests leads to a large amount of undefined symbols errors:

```
[100%] Linking CXX executable run_tests
Undefined symbols for architecture x86_64:
  "CppUnit::SourceLine::SourceLine(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)", referenced from:
      ModuleProfileTest::testGetName()       in ModuleProfileTest.cpp.o
      ModuleProfileTest::testGetDescription()       in ModuleProfileTest.cpp.o
      ModuleProfileTest::testGetContent()       in ModuleProfileTest.cpp.o
      ModulePackageTest::testSimpleGetters()       in ModulePackageTest.cpp.o
      ModulePackageTest::testGetArtifacts()       in ModulePackageTest.cpp.o
      ModulePackageTest::testGetProfiles()       in ModulePackageTest.cpp.o
      ModulePackageTest::testGetModuleDependencies()       in ModulePackageTest.cpp.o
      ...
  "CppUnit::TextTestRunner::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, bool, bool)", referenced from:
      _main in run_tests.cpp.o
  "CppUnit::AdditionalMessage::AdditionalMessage(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      void CppUnit::assertEquals<int>(int const&, int const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in DependencyContainerTest.cpp.o
      void CppUnit::assertEquals<int>(int const&, int const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in CompsEnvironmentItemTest.cpp.o
      void CppUnit::assertEquals<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in CompsEnvironmentItemTest.cpp.o
      void CppUnit::assertEquals<int>(int const&, int const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in CompsGroupItemTest.cpp.o
      void CppUnit::assertEquals<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in CompsGroupItemTest.cpp.o
      void CppUnit::assertEquals<int>(int const&, int const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in TransactionItemReasonTest.cpp.o
      void CppUnit::assertEquals<libdnf::TransactionItemReason>(libdnf::TransactionItemReason const&, libdnf::TransactionItemReason const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in TransactionItemReasonTest.cpp.o
      ...
  "CppUnit::CompilerOutputter::CompilerOutputter(CppUnit::TestResultCollector*, std::basic_ostream<char, std::char_traits<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      _main in run_tests.cpp.o
  "CppUnit::TestFactoryRegistry::getRegistry(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      __GLOBAL__sub_I_ModuleProfileTest.cpp in ModuleProfileTest.cpp.o
      __GLOBAL__sub_I_ModulePackageTest.cpp in ModulePackageTest.cpp.o
      __GLOBAL__sub_I_ContextTest.cpp in ContextTest.cpp.o
      __GLOBAL__sub_I_ModulePackageContainerTest.cpp in ModulePackageContainerTest.cpp.o
      __GLOBAL__sub_I_PackageTest.cpp in PackageTest.cpp.o
      __GLOBAL__sub_I_DependencyTest.cpp in DependencyTest.cpp.o
      __GLOBAL__sub_I_DependencyContainerTest.cpp in DependencyContainerTest.cpp.o
      ...
  "CppUnit::Message::addDetail(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      TransactionTest::testInsert()       in TransactionTest.cpp.o
      TransactionTest::testInsertWithSpecifiedId()       in TransactionTest.cpp.o
      TransactionTest::testInsert() [clone .cold] in TransactionTest.cpp.o
      TransactionTest::testInsertWithSpecifiedId() [clone .cold] in TransactionTest.cpp.o
  "CppUnit::Message::Message(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      TransactionTest::testInsert()       in TransactionTest.cpp.o
      TransactionTest::testInsertWithSpecifiedId()       in TransactionTest.cpp.o
  "CppUnit::Message::Message(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      ModuleProfileTest::testGetName()       in ModuleProfileTest.cpp.o
      ModuleProfileTest::testGetDescription()       in ModuleProfileTest.cpp.o
      ModuleProfileTest::testGetContent()       in ModuleProfileTest.cpp.o
      ModulePackageTest::testSimpleGetters()       in ModulePackageTest.cpp.o
      ModulePackageTest::testGetArtifacts()       in ModulePackageTest.cpp.o
      ModulePackageTest::testGetProfiles()       in ModulePackageTest.cpp.o
      ModulePackageTest::testGetModuleDependencies()       in ModulePackageTest.cpp.o
      ...
  "CppUnit::Message::Message(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      MergedTransactionTest::testMerge()      in MergedTransactionTest.cpp.o
  "CppUnit::Asserter::failNotEqual(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, CppUnit::SourceLine const&, CppUnit::AdditionalMessage const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)", referenced from:
      void CppUnit::assertEquals<int>(int const&, int const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in DependencyContainerTest.cpp.o
      void CppUnit::assertEquals<int>(int const&, int const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in CompsEnvironmentItemTest.cpp.o
      void CppUnit::assertEquals<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in CompsEnvironmentItemTest.cpp.o
      void CppUnit::assertEquals<int>(int const&, int const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in CompsGroupItemTest.cpp.o
      void CppUnit::assertEquals<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in CompsGroupItemTest.cpp.o
      void CppUnit::assertEquals<int>(int const&, int const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in TransactionItemReasonTest.cpp.o
      void CppUnit::assertEquals<libdnf::TransactionItemReason>(libdnf::TransactionItemReason const&, libdnf::TransactionItemReason const&, CppUnit::SourceLine, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in TransactionItemReasonTest.cpp.o
      ...
  "CppUnit::TestCase::TestCase(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)", referenced from:
      ModuleProfileTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in ModuleProfileTest.cpp.o
      ModulePackageTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in ModulePackageTest.cpp.o
      CppUnit::TestSuiteFactory<ContextTest>::makeTest()       in ContextTest.cpp.o
      ModulePackageContainerTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in ModulePackageContainerTest.cpp.o
      PackageTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in PackageTest.cpp.o
      DependencyTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in DependencyTest.cpp.o
      DependencyContainerTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in DependencyContainerTest.cpp.o
      ...
  "CppUnit::TestSuite::TestSuite(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)", referenced from:
      CppUnit::TestSuiteFactory<ModuleProfileTest>::makeTest()       in ModuleProfileTest.cpp.o
      CppUnit::TestSuiteFactory<ModulePackageTest>::makeTest()       in ModulePackageTest.cpp.o
      CppUnit::TestSuiteFactory<ContextTest>::makeTest()       in ContextTest.cpp.o
      CppUnit::TestSuiteFactory<ModulePackageContainerTest>::makeTest()       in ModulePackageContainerTest.cpp.o
      CppUnit::TestSuiteFactory<PackageTest>::makeTest()       in PackageTest.cpp.o
      CppUnit::TestSuiteFactory<DependencyTest>::makeTest()       in DependencyTest.cpp.o
      CppUnit::TestSuiteFactory<DependencyContainerTest>::makeTest()       in DependencyContainerTest.cpp.o
      ...
  "CppUnit::TestSuiteBuilderContextBase::getTestNameFor(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const", referenced from:
      ModuleProfileTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in ModuleProfileTest.cpp.o
      ModulePackageTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in ModulePackageTest.cpp.o
      CppUnit::TestSuiteFactory<ContextTest>::makeTest()       in ContextTest.cpp.o
      ModulePackageContainerTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in ModulePackageContainerTest.cpp.o
      PackageTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in PackageTest.cpp.o
      DependencyTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in DependencyTest.cpp.o
      DependencyContainerTest::addTestsToSuite(CppUnit::TestSuiteBuilderContextBase&)         in DependencyContainerTest.cpp.o
      ...
  "CppUnit::Test::findTestPath(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CppUnit::TestPath&) const", referenced from:
      vtable for CppUnit::TestCaller<ModuleProfileTest> in ModuleProfileTest.cpp.o
      vtable for ModuleProfileTest in ModuleProfileTest.cpp.o
      vtable for CppUnit::TestCaller<ModulePackageTest> in ModulePackageTest.cpp.o
      vtable for ModulePackageTest in ModulePackageTest.cpp.o
      vtable for CppUnit::TestCaller<ContextTest> in ContextTest.cpp.o
      vtable for ContextTest in ContextTest.cpp.o
      vtable for CppUnit::TestCaller<ModulePackageContainerTest> in ModulePackageContainerTest.cpp.o
      ...
  "CppUnit::Test::resolveTestPath(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const", referenced from:
      vtable for CppUnit::TestCaller<ModuleProfileTest> in ModuleProfileTest.cpp.o
      vtable for ModuleProfileTest in ModuleProfileTest.cpp.o
      vtable for CppUnit::TestCaller<ModulePackageTest> in ModulePackageTest.cpp.o
      vtable for ModulePackageTest in ModulePackageTest.cpp.o
      vtable for CppUnit::TestCaller<ContextTest> in ContextTest.cpp.o
      vtable for ContextTest in ContextTest.cpp.o
      vtable for CppUnit::TestCaller<ModulePackageContainerTest> in ModulePackageContainerTest.cpp.o
      ...
  "CppUnit::Test::findTest(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const", referenced from:
      vtable for CppUnit::TestCaller<ModuleProfileTest> in ModuleProfileTest.cpp.o
      vtable for ModuleProfileTest in ModuleProfileTest.cpp.o
      vtable for CppUnit::TestCaller<ModulePackageTest> in ModulePackageTest.cpp.o
      vtable for ModulePackageTest in ModulePackageTest.cpp.o
      vtable for CppUnit::TestCaller<ContextTest> in ContextTest.cpp.o
      vtable for ContextTest in ContextTest.cpp.o
      vtable for CppUnit::TestCaller<ModulePackageContainerTest> in ModulePackageContainerTest.cpp.o
      ...
  "CppUnit::TestCase::getName[abi:cxx11]() const", referenced from:
      vtable for CppUnit::TestCaller<ModuleProfileTest> in ModuleProfileTest.cpp.o
      vtable for ModuleProfileTest in ModuleProfileTest.cpp.o
      vtable for CppUnit::TestCaller<ModulePackageTest> in ModulePackageTest.cpp.o
      vtable for ModulePackageTest in ModulePackageTest.cpp.o
      vtable for CppUnit::TestCaller<ContextTest> in ContextTest.cpp.o
      vtable for ContextTest in ContextTest.cpp.o
      vtable for CppUnit::TestCaller<ModulePackageContainerTest> in ModulePackageContainerTest.cpp.o
      ...
  "CppUnit::TestNamer::getFixtureName[abi:cxx11]() const", referenced from:
      CppUnit::TestSuiteFactory<ModuleProfileTest>::makeTest()       in ModuleProfileTest.cpp.o
      CppUnit::TestSuiteFactory<ModulePackageTest>::makeTest()       in ModulePackageTest.cpp.o
      CppUnit::TestSuiteFactory<ContextTest>::makeTest()       in ContextTest.cpp.o
      CppUnit::TestSuiteFactory<ModulePackageContainerTest>::makeTest()       in ModulePackageContainerTest.cpp.o
      CppUnit::TestSuiteFactory<PackageTest>::makeTest()       in PackageTest.cpp.o
      CppUnit::TestSuiteFactory<DependencyTest>::makeTest()       in DependencyTest.cpp.o
      CppUnit::TestSuiteFactory<DependencyContainerTest>::makeTest()       in DependencyContainerTest.cpp.o
      ...
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [tests/CMakeFiles/run_tests.dir/build.make:368: tests/run_tests] Error 1
make[1]: *** [CMakeFiles/Makefile2:833: tests/CMakeFiles/run_tests.dir/all] Error 2
make: *** [Makefile:160: all] Error 2
```

**To Reproduce**
1. Remove the postpatch in pkgs/tools/package-management/libdnf/default.nix
2. Build
3. Watch the missing symobl errors

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
libdnf
```

### PR DESCRIPTION
**Pull request by katexochen**
## Description of changes

- Adding `dnf4`.
- Adding `libcomps`, which is a dependency of dnf4. libcomps is mostly C, but uses pyproject/setuptools for the python build, so I've placed it in `python3.pkgs`. Not sure this is right.
- Enabling python bindings in `libdnf`. The cppunit issue seems to have resolved itself, I've just re-added the dependency and removed the fix. Closes #107428.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by rb2k** [comment]
Oh neat!

I'm afraid I don't have access to our macs at work anymore (switched teams) which is where I usually tested the microdnf stuff.

Seems like the biuld logs at least on aarch64-darwin seem angry about:

`error: attribute 'tests' in selection path 'libdnf.passthru.tests' not found`

**Comment by katexochen** [comment]
@rb2k I'm still working on some things. Currently I'm trying to figure out how I can make the Python bindings of libdnf importable by Python, that doesn't seem to work yet. I'm also looking into the dnf4 plugins...

> `error: attribute 'tests' in selection path 'libdnf.passthru.tests' not found`

Not sure if this is a problem, the test is marked as skipped.

**Comment by rb2k** [comment]
> @rb2k I'm still working on some things. Currently I'm trying to figure out how I can make the Python bindings of libdnf importable by Python, that doesn't seem to work yet.

Yeah, the Python bindings and generation always caused me headaches. Before we went with microdnf, I tried to get 'proper' dnf to compile and ran into issues in that area. That being said, dnf4 might have less 'problem surface'. Fingers crossed :)

**Comment by katexochen** [comment]
`libdnf` and `libcomps` seem to be good now, both packaged as C and Python library. I finally understand multi output derivations. :D 

Now trying to get man pages working for dnf and looking into the plugins.

**Comment by katexochen** [comment]
dnf man pages should be working now, too. I will open a separate PR for the plugins.

**Review by malt3 — APPROVED** [review]

**Review by msanft — APPROVED** [review]
Built on `x86_64-linux`

**Comment by malt3** [comment]
Result of `nixpkgs-review pr 266635` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python310Packages.libdnf (python310Packages.libdnf.dev ,python310Packages.libdnf.py)</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>dnf4</li>
    <li>dnf4.man</li>
    <li>libcomps</li>
    <li>libcomps.dev</li>
    <li>libcomps.py (python311Packages.libcomps ,python311Packages.libcomps.dev ,python311Packages.libcomps.py)</li>
    <li>libdnf</li>
    <li>libdnf.dev</li>
    <li>libdnf.py (python311Packages.libdnf ,python311Packages.libdnf.dev ,python311Packages.libdnf.py)</li>
    <li>microdnf</li>
    <li>python310Packages.libcomps (python310Packages.libcomps.dev ,python310Packages.libcomps.py)</li>
  </ul>
</details>

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -6137,8 +6137,22 @@ self: super: with self; {
...
+  ];
+
   libcst = callPackage ../development/python-modules/libcst { };
 
+  libdnf = lib.pipe pkgs.libdnf [
+    toPythonModule
+    (p: p.overrideAttrs (super: { meta = super.meta // { outputsToInstall = [ "py" ]; }; }))
+    (p: p.override { inherit python; })
+    (p: p.py)
+  ];
```
Comment (malt3) [inline pkgs/top-level/python-packages.nix:6154]: This seems to always set python = python311

Comment (katexochen) [reply pkgs/top-level/python-packages.nix:6156]: Thanks for catching this, there was a patch missing in libdnf to fix the install path.

**Comment by katexochen** [comment]
Result of `nixpkgs-review pr 266635` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>dnf4 (python311Packages.dnf4)</li>
    <li>dnf4.man (python311Packages.dnf4.man)</li>
    <li>libcomps</li>
    <li>libcomps.dev</li>
    <li>libcomps.py (python311Packages.libcomps ,python311Packages.libcomps.dev ,python311Packages.libcomps.py)</li>
    <li>libdnf</li>
    <li>libdnf.dev</li>
    <li>libdnf.py (python311Packages.libdnf ,python311Packages.libdnf.dev ,python311Packages.libdnf.py)</li>
    <li>microdnf</li>
    <li>python310Packages.dnf4</li>
    <li>python310Packages.dnf4.man</li>
    <li>python310Packages.libcomps (python310Packages.libcomps.dev ,python310Packages.libcomps.py)</li>
    <li>python310Packages.libdnf (python310Packages.libdnf.dev ,python310Packages.libdnf.py)</li>
  </ul>
</details>

**Review by malt3 — APPROVED** [review]

File: pkgs/tools/package-management/libdnf/default.nix [hunk 2]
```diff
@@ -56,29 +67,34 @@ stdenv.mkDerivation rec {
...
       --replace '$'{prefix}/@CMAKE_INSTALL_LIBDIR@ @CMAKE_INSTALL_FULL_LIBDIR@
+    substituteInPlace cmake/modules/FindPythonInstDir.cmake \
+      --replace "@PYTHON_INSTALL_DIR@" "$out/${python.sitePackages}"
   '';
 
   cmakeFlags = [
     "-DWITH_GTKDOC=OFF"
     "-DWITH_HTML=OFF"
-    "-DWITH_BINDINGS=OFF"
+    "-DPYTHON_DESIRED=${lib.substring 0 1 python.version}"
```
Comment (pbsds) [inline pkgs/tools/package-management/libdnf/default.nix:85]:
```suggestion
    "-DPYTHON_DESIRED=${lib.head (lib.splitString ["."] python.version)}"
```

File: pkgs/development/python-modules/dnf4/default.nix [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  ];
+
+  propagatedBuildInputs = [
+    libcomps
+    libdnf
+    rpm
+  ];
+
+  cmakeFlags = [
+    "-DPYTHON_DESIRED=3"
```
Comment (pbsds) [inline pkgs/development/python-modules/dnf4/default.nix:57]: should this also depend on `python.version`, for consistency sake?

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -6137,8 +6139,22 @@ self: super: with self; {
 
   libcloud = callPackage ../development/python-modules/libcloud { };
 
+  libcomps = lib.pipe pkgs.libcomps [
+    toPythonModule
+    (p: p.overrideAttrs (super: { meta = super.meta // { outputsToInstall = [ "py" ]; }; }))
+    (p: p.override { inherit python; })
+    (p: p.py)
```
Comment (pbsds) [inline pkgs/top-level/python-packages.nix:6146]: the `.py` strikes me as yucky, but it is also incredibly creative and does reduce the install size. Note that the `propagatedBuildInputs` go to `dev` and not `py`. Luckily it seems these packages are unaffected

**Review by pbsds — COMMENTED** [review]
Result of `nixpkgs-review pr 266635` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>dnf4 (python311Packages.dnf4)</li>
    <li>dnf4.man (python311Packages.dnf4.man)</li>
    <li>libcomps</li>
    <li>libcomps.dev</li>
    <li>libcomps.py (python311Packages.libcomps ,python311Packages.libcomps.dev ,python311Packages.libcomps.py)</li>
    <li>libdnf</li>
    <li>libdnf.dev</li>
    <li>libdnf.py (python311Packages.libdnf ,python311Packages.libdnf.dev ,python311Packages.libdnf.py)</li>
    <li>microdnf</li>
    <li>python310Packages.dnf4</li>
    <li>python310Packages.dnf4.man</li>
    <li>python310Packages.libcomps (python310Packages.libcomps.dev ,python310Packages.libcomps.py)</li>
    <li>python310Packages.libdnf (python310Packages.libdnf.dev ,python310Packages.libdnf.py)</li>
  </ul>
</details>


---

- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on ARCHITECTURE
- [x] executables tested on ARCHITECTURE
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [ ] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

---

I see sphinx is included, but not html docs are produced. Is sphinx strictly neccesary?

Aside from my general desire to get rid of sphinx from the build closure, this mostly LGTM. If sphinx is to be kept, please move it to the nativeBuildInputs

**Comment by pbsds** [comment]
Also, these two binaries seem to not find a link target:

```shell
$ fd --no-ignore --hidden -L . results/    --type x -x bash -c 'ldd {} | grep "not found" -q && echo {}' 2>/dev/null
results/libdnf.py/lib/python3.11/site-packages/hawkey/test/_hawkey_test.so
results/python310Packages.libdnf/lib/python3.10/site-packages/hawkey/test/_hawkey_test.so
```

Are they even needed?

**Comment by katexochen** [comment]
> I see sphinx is included, but not html docs are produced. Is sphinx strictly neccesary?

Sphinx is used for generating the man pages, see
https://github.com/rpm-software-management/dnf/blob/cb54681d17cbd83e28261b1481e4945d2b7f494c/doc/CMakeLists.txt#L15-L18

File: pkgs/top-level/python-packages.nix (hunk 4 above)
Comment (pbsds) [reply pkgs/top-level/python-packages.nix:6146]: it's fine really. i just hadn't seen it before

**Pull request merged**